File: 3BEG - Robot Testing Factory Factory/robotPaths.hs
module Main where

import Control.Monad

data Robot = Robot {x :: Int ,
                    y :: Int ,
                    dir :: Int,
                    rotation :: Int } deriving (Show)
                                               
main :: IO () 
main = do
  tStr <- getLine
  let numTests = read tStr :: Int
  testStrs <- replicateM numTests getLine
  let results = [(run (Robot x y dir rot) t w h) | 
                  line <- map words testStrs, 
                  let [w, h, x, y, dir, rot, t] = map (\s -> read s::Int) line] 
  mapM_ (\(x, y) -> putStrLn (show x ++ " " ++ show y)) results

run :: Robot -> Int -> Int -> Int -> (Int, Int)
run (Robot x y _ _) 0 _ _ = (x,y)
run (Robot 0 0  _ _) _ 1 1 = (0,0)
run robot t w h = run (nextMove robot) (t-1) w h 
  where
    nextMove :: Robot -> Robot  
    nextMove r = 
      if checkMove w h $ move r
      then move r   
      else nextMove $ rotate r   
  
move :: Robot -> Robot
move (Robot x y dir rot) = 
  case dir of
    90  -> Robot x (y-1) dir rot
    270 -> Robot x (y+1) dir rot
    180 -> Robot (x-1) y dir rot
    0   -> Robot (x+1) y dir rot
    
rotate :: Robot ->  Robot
rotate (Robot x y dir rot) = Robot x y (fixDir $ dir+rot) rot where
  fixDir dir | dir >= 360 = dir - 360 
             | dir < 0   = dir + 360
             | otherwise = dir              

checkMove :: Int -> Int -> Robot -> Bool
checkMove w h (Robot x y _ _) = (x >= 0 && x < w) && (y >= 0 && y < h) 

File: 3BEG - Robot Testing Factory Factory/robotPaths.cpp

#include <iostream>
#include <stdlib.h>
using namespace std;

class Bot {
  public: 
    int x, y, dir, rot;
    void rotate();
    Bot move();
};

bool isValid(Bot, int, int);

int main() {
  int numTests, w, h, t;
  Bot bot;
  
  cin >> numTests;
  for(int i=0; i < numTests; i++){
    cin >> w >> h >> bot.x >> bot.y >> bot.dir >> bot.rot >> t;
    if (w > 1 && h > 1)
    {
      while(t > 0)
      {
        if (isValid(bot.move(), w, h))
	{
          bot = bot.move();
          t--;
	}
        else
          bot.rotate();
      }
    }
    cout << bot.x << " " << bot.y << endl;
  }    
  return 0;
}

void Bot::rotate()
{
  dir += rot;
  if (dir < 0)
    dir += 360;
  else if (dir >= 360)
    dir -= 360;
}

Bot Bot::move()
{
  Bot temp;
  temp.x = x; temp.y = y; temp.dir = dir; temp.rot = rot;
  switch (temp.dir){
  case 0: temp.x++; break;
  case 90: temp.y--; break;
  case 180: temp.x--; break;
  case 270: temp.y++; break;
  default: cerr << "HOW DID THIS HAPPEN?!?!?!\n" << temp.dir; exit(1); 
  }
  return temp;
}

bool isValid(Bot b, int w, int h)
{
  return (b.x >= 0 && b.x < w) && (b.y >= 0 && b.y < h);
}
